hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/SmallFileDataStreamOutput.java

```diff
@@ -0,0 +1,564 @@
+  private OmKeyLocationInfo keyLocationInfo;
+  private final OmKeyArgs keyArgs;
+
+  private DataStreamOutput dataStreamOutput;
+
+  private boolean unsafeByteBufferConversion;
+
+  private final ByteBuffer currentBuffer;
+  private long versionID;
+  private final Token<OzoneBlockTokenIdentifier> token;
+
+  // error handler
+  private final ExcludeList excludeList;
+  private final Map<Class<? extends Throwable>, RetryPolicy> retryPolicyMap;
+  private int retryCount;
+
+  public SmallFileDataStreamOutput(
+      OpenKeySession handler,
+      XceiverClientFactory xceiverClientManager,
+      OzoneManagerProtocol omClient,
+      OzoneClientConfig config,
+      boolean unsafeByteBufferConversion
+  ) throws IOException {
+    this.xceiverClientFactory = xceiverClientManager;
+    this.omClient = omClient;
+    this.config = config;
+    this.openKeySession = handler;
+
+    this.keyLocationInfo = handler.getKeyInfo().getLatestVersionLocations()
+        .getLocationList(handler.getOpenVersion()).get(0);
+    this.blockID = new AtomicReference<>(keyLocationInfo.getBlockID());
+    this.versionID = keyLocationInfo.getCreateVersion();
+
+    this.unsafeByteBufferConversion = unsafeByteBufferConversion;
+
+    OmKeyInfo info = handler.getKeyInfo();
+    this.currentBuffer = ByteBuffer.allocate((int) info.getDataSize());
+
+    this.keyArgs = new OmKeyArgs.Builder().setVolumeName(info.getVolumeName())
+        .setBucketName(info.getBucketName()).setKeyName(info.getKeyName())
+        .setReplicationConfig(info.getReplicationConfig())
+        .setDataSize(info.getDataSize())
+        .setIsMultipartKey(false).build();
+
+    this.retryPolicyMap = HddsClientUtils.getRetryPolicyByException(
+        config.getMaxRetryCount(), config.getRetryInterval());
+    this.retryCount = 0;
+
+    this.excludeList = new ExcludeList();
+
+    this.token = null;
+  }
+
+
+  @VisibleForTesting
+  public BlockID getBlockID() {
+    return blockID.get();
+  }
+
+  @VisibleForTesting
+  public OmKeyLocationInfo getKeyLocationInfo() {
+    return keyLocationInfo;
+  }
+
+  private OmKeyLocationInfo allocateNewBlock() throws IOException {
+    if (!excludeList.isEmpty()) {
+      LOG.info("Allocating block with {}", excludeList);
+    }
+    OmKeyLocationInfo omKeyLocationInfo =
+        omClient.allocateBlock(keyArgs, openKeySession.getId(), excludeList);
+
+    this.keyLocationInfo = omKeyLocationInfo;
+    this.blockID.set(keyLocationInfo.getBlockID());
+    this.versionID = keyLocationInfo.getCreateVersion();
+
+    return omKeyLocationInfo;
+  }
+
+  @Override
+  public void write(ByteBuffer bb) throws IOException {
+    if (bb == null) {
+      throw new NullPointerException();
+    }
+
+    byte b = bb.array()[0];
+    currentBuffer.put(b);
+  }
+
+  @Override
+  public void write(ByteBuffer bb, int off, int len) throws IOException {
+    if (bb == null) {
+      throw new NullPointerException();
+    }
+
+    byte[] b = bb.array();
```

Comment:
We cannot call array() since the ByteBuffer may not have an array.

Also, using an array means there must be buffer copying inside the code.  This is the reason that we use ByteBuffer but not byte[].